index.js

```diff
@@ -1,47 +1,136 @@
+  self._elem = elem
+  self._mediaSource = mediaSource
+  self._openHandler = self._onSourceOpen.bind(self, obj)
+  self._flowHandler = self._flow.bind(self)
+  self._bufferDuration = opts.bufferDuration || DEFAULT_BUFFER_DURATION
 
-  self._mediaSource.addEventListener('sourceopen', function () {
-    if (MediaSource.isTypeSupported(self._type)) {
-      self._sourceBuffer = self._mediaSource.addSourceBuffer(self._type)
-      self._sourceBuffer.addEventListener('updateend', self._flow.bind(self))
-      self._flow()
+  if (typeof obj === 'string') {
+    // Need to create a new sourceBuffer
+    if (self._mediaSource.readyState === 'open') {
+      self._createSourceBuffer(obj)
     } else {
-      self._mediaSource.endOfStream('decode')
+      self._mediaSource.addEventListener('sourceopen', self._openHandler)
     }
+  } else if (obj._sourceBuffer) {
+    obj.destroy()
+    self._sourceBuffer = obj._sourceBuffer // Copy over the old sourceBuffer
+    self._sourceBuffer.addEventListener('updateend', self._flowHandler)
+  } else {
+    throw new Error('The argument to MediaElementWrapper.getStream must be a string or a previous stream returned from that function')
+  }
+
+  self._elem.addEventListener('timeupdate', self._flowHandler)
+
+  self.on('error', function () {
+    try {
+      self._mediaSource.endOfStream('decode')
+    } catch (err) {}
   })
 
   self.on('finish', function () {
-    self._mediaSource.endOfStream()
+    if (self.destroyed) return
+    self._finished = true
+    if (self._allStreams.every(function (other) { return other._finished })) {
+      self._mediaSource.endOfStream()
+    }
   })
 }
 
+MediaSourceStream.prototype._onSourceOpen = function (type) {
+  var self = this
+  if (self.destroyed) return
+
+  self._mediaSource.removeEventListener('sourceopen', self._openHandler)
+  self._createSourceBuffer(type)
+}
+
+MediaSourceStream.prototype.destroy = function (err) {
+  var self = this
+  if (self.destroyed) return
+  self.destroyed = true
+
+  // Remove from allStreams
+  self._allStreams.splice(self._allStreams.indexOf(self), 1)
+
+  self._mediaSource.removeEventListener('sourceopen', self._openHandler)
+  self._elem.removeEventListener('timeupdate', self._flowHandler)
+  if (self._sourceBuffer) {
+    self._sourceBuffer.removeEventListener('updateend', self._flowHandler)
+    if (self._mediaSource.readyState === 'open') {
+      self._sourceBuffer.abort()
+    }
+  }
+
+  if (err) self.emit('error', err)
+  self.emit('close')
+}
+
+MediaSourceStream.prototype._createSourceBuffer = function (type) {
+  var self = this
+  if (self.destroyed) return
+
+  if (MediaSource.isTypeSupported(type)) {
+    self._sourceBuffer = self._mediaSource.addSourceBuffer(type)
+    self._sourceBuffer.addEventListener('updateend', self._flowHandler)
+    if (self._cb) {
+      var cb = self._cb
+      self._cb = null
+      cb()
+    }
+  } else {
+    self.emit('error', new Error('The provided type is not supported'))
```

Comment:
This should call `self.destroy(err)` for the same reason as mentioned in another comment. We want `'close'` to get emitted so that `unpipe` gets called: https://github.com/nodejs/readable-stream/blob/master/lib/_stream_readable.js#L618-L622